File: Data Structures & Algorithms/count-number-of-islands/submission-2.cpp
class Solution {
public:
    int direction[4][2]={{1,0},{-1,0},{0,1},{0,-1}};
    int numIslands(vector<vector<char>>& grid) {
        int row=grid.size(), col=grid[0].size();
        int island=0;
        for(int r=0;r<row;r++){
            for(int c=0;c<col;c++){
                if(grid[r][c]=='1'){
                    dfs(grid,r,c);
                    island++;
                }
            }
        }
        return island;
    }
    void dfs(vector<vector<char>>&grid, int r, int c){
        if(r<0|| c<0|| r>=grid.size()|| c>=grid[0].size() || grid[r][c]=='0'){
            return;
        }
        grid[r][c]='0';
        for(int i=0;i<4;i++){
            dfs(grid,r+direction[i][0],c+direction[i][1]);
        }
    }
};

File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    int[] dp;
    public int climbStairs(int n) {
        dp= new int[n+1];
        Arrays.fill(dp, -1);
        return dfs(n,0);
    }
    public int dfs(int n, int i){
        if(i>=n){
            return i== n?1:0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        return dp[i]=dfs(n, i+1)+ dfs(n, i+2);
    }
}
